guides/source/active_record_callbacks.md

```diff
@@ -329,124 +763,267 @@ Just as with validations, it is also possible to skip callbacks by using the fol
+during seeding the database with test data.
 
 ```ruby
 class User < ApplicationRecord
-  has_many :articles, dependent: :destroy
+  after_create :send_welcome_email
+
+  def send_welcome_email
+    puts "Welcome email sent to #{self.email}"
+  end
 end
+```
 
-class Article < ApplicationRecord
-  after_destroy :log_destroy_action
+In this example, the `after_create` callback triggers the `send_welcome_email`
+method every time a new user is created.
 
-  def log_destroy_action
-    puts 'Article destroyed'
+To create a user without sending the welcome email, we can use the
+`ActiveRecord::Suppressor` module as follows:
+
+```ruby
+User.suppress do
+  User.create(name: "Jane", email: "jane@example.com")
+end
+```
+
+In the above code, the `User.suppress` block ensures that the
+`send_welcome_email` callback is not executed during the creation of the "Jane"
+user, allowing us to create the user without sending the welcome email.
+
+WARNING: Using the Active Record Suppressor, while potentially beneficial for
+selectively controlling callback execution, can introduce complexity and
+unexpected behavior. Suppressing callbacks can obscure the intended flow of your
+application, leading to difficulties in understanding and maintaining the
+codebase over time. Carefully consider the implications of suppressing
+callbacks, ensuring thorough documentation and thoughtful testing to mitigate
+risks of unintended side effects, performance issues, and test failures.
+
+Halting Execution
+-----------------
+
+As you start registering new callbacks for your models, they will be queued for
+execution. This queue will include all of your model's validations, the
+registered callbacks, and the database operation to be executed.
+
+The whole callback chain is wrapped in a transaction. If any callback raises an
+exception, the execution chain gets halted and a **rollback** is issued, and the
+error will be re-raised.
+
+```ruby
+class Product < ActiveRecord::Base
+  before_validation do
+    raise "Price can't be negative" if total_price < 0
   end
 end
+
+Product.create # raises "Price can't be negative"
 ```
 
-```irb
-irb> user = User.first
-=> #<User id: 1>
-irb> user.articles.create!
-=> #<Article id: 1, user_id: 1>
-irb> user.destroy
-Article destroyed
-=> #<User id: 1>
+This unexpectedly breaks code that does not expect methods like `create` and
+`save` to raise exceptions.
+
+Instead, you should use `throw :abort` to intentionally halt the chain. If any
+callback throws `:abort`, the process will be aborted and `create` will return
+false.
+
+```ruby
+class Product < ActiveRecord::Base
+  before_validation do
+    throw :abort if total_price < 0
+  end
+end
+
+Product.create # => false
 ```
 
+WARNING: If an exception occurs during the callback chain, Rails will re-raise
+it unless it is an `ActiveRecord::Rollback`, `ActiveRecord::RecordNotSaved` or
+`ActiveRecord::RecordInvalid` exception.
```

Comment:
If we move the WARNING higher up, I think it would be nice to describe `create!` raising as well.
This would make it similar to `destroy!` below.
````suggestion
However, it will raise an `ActiveRecord::RecordNotSaved` when calling
`create!`. This exception indicates that the record was not saved due to
the callback's interruption.

```ruby
User.create! # => raises an ActiveRecord::RecordNotSaved
```
````
```

Reply:
I like this, thank you!